src/main/java/org/skriptlang/skript/lang/script/Annotation.java

```diff
@@ -0,0 +1,183 @@
+ * By itself, it does nothing. There are no errors or penalties for writing a bogus
+ * or unknown annotation (in this case it acts like a comment does).
+ * An annotation is visible to (and only to) the line of code it is placed before:
+ *
+ * <pre>{@code
+ * 	on event:
+ *      @my cool annotation
+ * 		my annotated effect
+ * }</pre>
+ * <p>
+ * If multiple annotations are placed before a line of code then all will be visible
+ * to that line.
+ *
+ * <pre>{@code
+ * 	on event:
+ *      @my cool annotation
+ *      @another annotation
+ * 		my annotated effect
+ * }</pre>
+ * <p>
+ * Annotations are visible to every element within the line, and are discarded after the line has been
+ * parsed.
+ *
+ * <pre>{@code
+ * 	on event:
+ *      @my cool annotation
+ * 		my annotated effect
+ * 		my not-so-annotated effect
+ * }</pre>
+ * <p>
+ * It is up to elements within a line how to (or whether to) interact with annotations.
+ * Some may have a global effect on any syntax (e.g. suppressing a warning, changing parsing order)
+ * whereas others might be relevant ONLY to one syntax.
+ * <p>
+ * This interface represents a post-parsing annotation.
+ */
+public interface Annotation extends CharSequence {
+
+	static @NotNull Annotation create(String text) {
+		return new SimpleAnnotation(text);
+	}
+
+	static boolean isAnnotationPresent(String text) {
+		return ParserInstance.get().hasAnnotation(text);
+	}
+
+	static boolean isAnnotationPresent(SkriptPattern pattern) {
+		return ParserInstance.get().hasAnnotationMatching(pattern);
+	}
+
+	/**
+	 * @return The annotation's content after the initial '@'
+	 */
+	@NotNull
+	String value();
+
+	/**
+	 * Whether the content of this annotation equals the given value.
+	 */
+	default boolean valueEquals(@Nullable String value) {
+		return this.value().equals(value);
+	}
+
+	class Match implements Annotation {
+
+		private final Annotation annotation;
+		private final MatchResult result;
+
+		public Match(Annotation annotation, MatchResult result) {
+			this.annotation = annotation;
+			this.result = result;
+		}
+
+		public MatchResult result() {
+			return result;
+		}
+
+		public Annotation annotation() {
+			return annotation;
+		}
+
+		@Override
+		public @NotNull String value() {
+			return annotation.value();
+		}
+
+		@Override
+		public int length() {
+			return annotation.length();
+		}
+
+		@Override
+		public char charAt(int index) {
+			return annotation.charAt(index);
+		}
+
+		@NotNull
+		@Override
+		public CharSequence subSequence(int start, int end) {
+			return annotation.subSequence(start, end);
+		}
+
+		@Override
+		public String toString() {
+			return annotation.toString();
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (obj instanceof Annotation)
+				return Objects.equals(this.value(), ((Annotation) obj).value())
+					|| annotation.equals(obj);
+			else
+				return false;
+		}
+
+	}
+
+}
+
+final class SimpleAnnotation implements Annotation, CharSequence {
```

Comment:
```suggestion
final class SimpleAnnotation implements Annotation {
```
is it necessary to specify that it implements CharSequence since Annotation extends it?

Reply:
Only for clarity I suppose, I often include them anyway because it might not be immediately clear that an annotation is a charsequence.